pkg/command/command.go

```diff
@@ -48,7 +48,14 @@ func New(cmd ...string) *Command {
 // New creates a new command from the provided workDir and the command
 // arguments.
 func NewWithWorkDir(workDir string, cmd ...string) *Command {
-	args := strings.Fields(strings.Join(cmd, " "))
+	joined := strings.Join(cmd, " ")
+	args := strings.Fields(joined)
+
+	// Pipes will be handled in a dedicated sub shell to avoid handling
+	// all the readers and writers of the pipes
+	if strings.Contains(joined, "|") {
+		args = []string{"sh", "-c", joined}
+	}
 
 	command := func() *Command {
 		if len(args) == 0 {
```

Comment:
I am concerned with all this magic that is happening in the command package. Currently e.g. `command.New("echo", "foo", "|bar").Run()` will not do what I'd (and my guess is: most of the other users) expect.

I think we should not do this (first joining all the arguments, then splitting them up again, doing something special when we find that character, maybe something else when we find a different one, ...). This weirdness reminds me of a lot of weirdness in the current bash code, where for certain things you need to exactly know how to escape a certain character/whitespace to do what it normally does. Let's not start doing that here!

Let's please not mess with the arguments and by that push it to the callers to quote stuff "correctly" and figure out which magic is going on in this package, let's pass the arguments down to the `exec.Command` and the OS as they are, please let's not mess with them.

To figure out how we can go about that:
- What do we miss when we'd just pass the arguments through to the `exec.Command` or used the `exec` package without a wrapper directly?
- Would it help if we'd use e.g. https://github.com/b4b4r07/go-pipe or such for making pipelines easier to build on the user's side?

Reply:
I agree that it’s not the technically best way to achieve the goal. I played around with `io.Pipe()` to have something like:

```golang
    c1 := exec.Command("ls")
    c2 := exec.Command("wc", "-l")

    r, w := io.Pipe() 
    c1.Stdout = w
    c2.Stdin = r

    var b2 bytes.Buffer
    c2.Stdout = &b2

    c1.Start()
    c2.Start()
    c1.Wait()
    w.Close()
    c2.Wait()
    io.Copy(os.Stdout, &b2)
```
Then the API could look like:
```
status, err := command.New("ls").Pipe("wc -l").Run()
```
Looks good, but the implementation effort is tremendously. 😁

I’ll invest more time in finding a solution around this. 🤔

Reply:
For this specific case, https://github.com/b4b4r07/go-pipe might already be a solution? I am however not sure if that should be part of the command package?

And re: joining/splitting:
Can we please not do that and pass the arguments along as we get them?

I don't understand why we do that

https://github.com/kubernetes/release/blob/19e07342ae1f9c110adb32f7ea1d2721b1e76919/pkg/command/command.go#L51-L60

and not just 

```go
command := exec.Command(cmd[0], cmd[1:]...)
```
?

Is there a reason for joining & splitting and the anon func I don't understand/see?

Reply:
The main reason is to provide an easy to read interface, for example:
`command.Execute("echo hi")`
behaves like the same as:
`command.Execute("echo", "hi")`. 

The reason is that I’m now able to easily pull in variables without additional formatting, like:
`command.Execute("echo", myVar)`. 

Are you suggesting in splitting up the command and args? I think that would be fine. :)

Or do we want to avoid the splitting at all and stick to only the syntax:
`command.Execute("echo", "hi")`
I would be fine with that, but it increases the verbosity of the code.

Reply:
> The main reason is to provide an easy to read interface, for example:
> `command.Execute("echo hi")`

I think that might only be "easy" on a first glance. How would I handle anything with a white-space (a command, an argument, a filename, ...).

> Or do we want to avoid the splitting at all and stick to only the syntax:
> `command.Execute("echo", "hi")`

Yes. I'd prefer us to only do/support that. I wouldn't want us to do things like `args := strings.Fields(strings.Join(cmd, " "))`. Splitting with `strings.Fields` is wrong in this context (consider a element of your `cmd` slice having a whitespace in it).
Let's just stay safe and don't mess with the arguments, let's pass them on as we get them in.

> Are you suggesting in splitting up the command and args? I think that would be fine. :)

To keep it similar with `exec.Command` and to following the principle of least surprise, I'd slightly prefer that, but that's just me. Totally fine if you decide to not separate them. :man_shrugging: